client/src/store.ts

```diff
@@ -0,0 +1,61 @@
+import React, { useReducer, createContext, useContext } from 'react';
+import cloneDeep from 'lodash.clonedeep';
+
+const NODE_ENV = process.env.NODE_ENV || 'development';
+
+type State = {
+  apiUrl: string;
+};
+
+type Action = { type: 'action1' } | { type: 'action2' };
+
+type ActionDictionary = {
+  [key: string]: (state: State, action: Action) => State;
+};
+
+type StoreContext = { state: State; dispatch: React.Dispatch<Action> };
+
+// Define initial application state
+const initialState: State = {
+  apiUrl: {
+    development: 'http://127.0.0.1:8000/api/v1/',
+    production: new URL('/api/v1/', window.location.origin).href,
+    test: 'https://tutv-dev.herokuapp.com/api/v1/',
+  }[NODE_ENV],
+};
+
+// Create the context
+const StoreContext = createContext({} as StoreContext);
+
+// In order to define an action, first define it in the Action type definition above, then add a
+// handler function here.
+// A handler function should take state as input, and return a copy of the state object mutated
+const actionHandlers: ActionDictionary = {
+  action1(state: State): State {
+    console.log('ACTION 1');
+    return state;
+  },
+  action2(state: State): State {
+    console.log('ACTION 2');
+    return state;
+  },
+};
+
+const reducer = (state: State, action: Action): State => {
+  const actionHandler = actionHandlers[action.type];
+  const state2 = cloneDeep(state);
+  return actionHandler(state2, action);
+};
+
+// The StoreProvider component wraps the whole app so that children can use useStore, which
+// leverages the context API
+export const StoreProvider: React.FC = ({ children }) => {
+  const [state, dispatch] = useReducer(reducer, initialState);
+  return React.createElement(
```

Comment:
Why not JSX?

Reply:
No real reason, I guess because I didn't want to change the file from `ts` to `tsx` for this one line of code, and also because I feel like the StoreProvider works more explicitly as a HOC when children are passed as an argument to `createElement` instead of just put there in JSX

Reply:
In this case, StoreProvider isn't a HOC per se. If you would like it to be an HOC, make sure you follow [the conventions](https://reactjs.org/docs/higher-order-components.html). If you'd rather use it as a normal component (like here), let's use JSX.

Reply:
I'll change it!